gc.c

```diff
@@ -1901,23 +1956,46 @@ class_object_id(VALUE klass)
 }
 
 static VALUE
-object_id0(VALUE obj)
+object_id0(VALUE obj, bool shareable)
 {
     VALUE id = Qfalse;
 
-    if (rb_shape_has_object_id(rb_obj_shape(obj))) {
-        shape_id_t object_id_shape_id = rb_shape_transition_object_id(obj);
-        id = rb_obj_field_get(obj, object_id_shape_id);
+    rb_shape_t *current_shape = rb_obj_shape(obj);
+
+    if (rb_shape_has_object_id(current_shape)) {
+        shape_id_t object_id_shape_id = rb_shape_object_id(obj);
+
+        if (RB_UNLIKELY(shareable && RSHAPE(object_id_shape_id)->type == SHAPE_EXTERNAL_OBJ_ID)) {
+            RUBY_ASSERT(obj_to_id_tbl);
+            st_lookup(obj_to_id_tbl, obj, &id);
+        }
+        else {
+            id = rb_obj_field_get(obj, object_id_shape_id);
+        }
         RUBY_ASSERT(id, "object_id missing");
         return id;
     }
 
-    // rb_shape_object_id_shape may lock if the current shape has
-    // multiple children.
-    shape_id_t object_id_shape_id = rb_shape_transition_object_id(obj);
-
     id = generate_next_object_id();
-    rb_obj_field_set(obj, object_id_shape_id, id);
+    // If the object is shareable and doesn't have free capacity we can't safely
+    // resize it. So we have to store the object_id externally.
+    if (shareable && current_shape->next_field_index == current_shape->capacity) {
+        shape_id_t object_id_shape_id = rb_shape_transition_external_object_id(obj);
+
+        if (RB_UNLIKELY(!obj_to_id_tbl)) {
+            obj_to_id_tbl = st_init_numtable();
+            obj_to_id_value = TypedData_Wrap_Struct(0, &obj_to_id_tbl_type, obj_to_id_tbl);
+        }
+        st_insert(obj_to_id_tbl, obj, id);
+        rb_shape_set_shape_id(obj, object_id_shape_id);
```

Comment:
We're running into the same issue than https://github.com/ruby/ruby/pull/13289. If we on 32bits the shape_id is in the generic_ivar table...